### ISSUE TITLE
How to deal with decrementing a counter?

### ISSUE THREAD

**Issue by tobz** [brayniac/tic#71]
It kind of just struck me that there's no good way to increment/decrement a counter or a gauge!

Was curious if you had any thoughts on those types of operations.  Things that jumped to mind:

- I could just deal with it, and have two metrics -- one for increments and one for decrements -- and I could just grab the difference between them to simulate my aggregate value.
- Turn `Sample` into an enum so it can have variants that drive the actions of different data types when they're given a sample to integrate.  The performance of matching the variants might suck though.
- Another field in `Sample` that represents a delta to be applied to counters/gauges.  This would require  logic to be sprinkled in to do things like "if `delta` field is non-zero, add it to the value, otherwise, take the value of `count`".

The `Sample`-based changes would also require a decision about negative numbers, and the `delta`-based approach is even trickier, I think?, since it would implicitly be signed, and then you can't easily do something like a `saturating_sub` even if you _do_ decide that you're OK with not allowing negative values.

tl;dr: have you ever considering increment/decrement options in tic before? :)

**Comment by brayniac** [comment]
Trying to understand the use-case better. 

When would you decrement a counter? Can events "un-happen"? I suspect that the counters should remain monotonically increasing. If we track both enqueues and dequeues, we can use both stats to calculate the current queue depth. This allows us to track enqueue rate - so we know how much work we're getting, the dequeue rate - so we know how much work we're accomplishing, and through subtracting - we get the queue depth at the end of the interval. These would be expressed as `rate(enqueue)`, `rate(dequeue)`, and `enqueue - dequeue` within the monitoring query language.

For gauges, it looks like the gauge is explicitly setting the value of the gauge? I'd expect you can just send the new lower value. Here we may opt to allow `i64` if we need to represent negative values - but this will break symmetry with the "value histograms" - as the `histogram` crate operates on `u64` values.

I might be missing something to your use-case. Please let me know.

**Comment by tobz** [comment]
Sorry, yeah, I should have been more clear that I was trying to brainstorm rather than ask precisely how to mold `tic` to fit this use case.

I was thinking of a classical "clients connected" style of metric.. where the count goes up and down but is a single value.

I hadn't actually considered the granularity I'd get by tracking the "enqueue" and "dequeue" portions, so I think I'll just go with that approach. :)
